BinaryEncoding.md

```diff
@@ -92,42 +95,46 @@ for all sections. The encoding of all sections begins as follows:
 Each section is optional and may appear at most once.
 Known sections (from this list) may not appear out of order.
 
-* [Signatures](#signatures-section) section
-* [Import Table](#import-table-section) section
-* [Function Signatures](#function-signatures-section) section
-* [Indirect Function Table](#indirect-function-table-section) section
+* [Type](#type-section) section
+* [Import](#import-section) section
+* [Function](#function-section) section
+* [Indirection](#indirection-section) section
 * [Memory](#memory-section) section
-* [Export Table](#export-table-section) section
-* [Start Function](#start-function-section) section
-* [Function Bodies](#function-bodies-section) section
-* [Data Segments](#data-segments-section) section
+* [Export](#export-section) section
+* [Start](#start-section) section
+* [Code](#code-section) section
+* [Data](#data-section) section
 * [Names](#names-section) section
 
 The end of the last present section must coincide with the last byte of the
 module. The shortest valid module is 8 bytes (`magic number`, `version`,
 followed by zero sections).
 
-### Signatures section
+### Type section
 
-ID: `signatures`
+ID: `types`
 
-The signatures section declares all function signatures that will be used in the module.
+The type section declares all function signatures that will be used in the module.
 
 | Field | Type | Description |
 | ----- |  ----- | ----- |
-| count | `varuint32` | count of signature entries to follow |
-| entries | `signature_entry*` | repeated signature entries as described below |
+| count | `varuint32` | count of type entries to follow |
+| entries | `type_entry*` | repeated type entries as described below |
 
-#### Signature entry
+#### Type entry
 | Field | Type | Description |
 | ----- |  ----- | ----- |
+| form | `uint8` | `0x40`, indicating a function type |
```

Comment:
>
> What I don't understand is why all those future new things can't just be
> new forms of entries in the types section such that a type-id is all you
> need?
>

It actually bakes in a two-way restriction: all _non-nullary_ constructors
_must_ be named, and all _nullary_ constructors _cannot_ be named. My
intuition is that both these restrictions may turn out to be cumbersome or
even problematic in the future. At least it's not obvious to me that they
won't. For example, when such types need to cross module boundaries: either
as type imports/exports (you may need to export both kinds), or occurring
in the signature of some function (you may want to avoid exporting an
auxiliary type definition for every type you mention somewhere).

In contrast, where do you see the downside of avoiding conflicts between
the spaces until we understand the future better?

I'd actually invert that scheme, to avoid the negation.
>
> Wouldn't that give all the primitive types today (i32, etc) negative
> indices?
>

Yes, but would there be a problem with that?